src/Components/Components/src/RenderTree/RenderTreeFrame.cs

```diff
@@ -184,66 +204,72 @@ namespace Microsoft.AspNetCore.Components.RenderTree
         /// </summary>
         [FieldOffset(16)] public readonly string MarkupContent;
 
-        private RenderTreeFrame(int sequence, string elementName, int elementSubtreeLength)
+        // Element constructor
+        private RenderTreeFrame(int sequence, int elementSubtreeLength, string elementName, object elementKey)
             : this()
```

Comment:
Is there a particular purpose here to calling `this()`?

Reply:
If you don't, the compiler gets mad because there are unassigned readonly fields.

This whole approach of having multiple fields sharing memory slots, and having the whole struct be readonly, and the fact constructors can only be differentiated by signature and not by any other kind of name or identifier, hits compiler limitations all over the place. Something like C's `union` would be a more natural way to expressive this.

I think the low-levelness of this struct within the rendering system justifies going to all these lengths but I wouldn't do it in less perf-critical places.